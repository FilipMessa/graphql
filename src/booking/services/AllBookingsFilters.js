// @flow

import idx from 'idx';
import { DateTime } from 'luxon';

import type { BookingsItem } from '../Booking';
import type { Leg } from '../../flight/Flight';
import { CoveredBy } from '../../flight/types/enums/CoveredBy';

export function groupBookings(bookings: $ReadOnlyArray<BookingsItem>) {
  return bookings.reduce(
    (acc, curVal) => {
      if (isPastBooking(curVal) === true) {
        acc.past.push(curVal);
      } else {
        acc.future.push(curVal);
      }
      return acc;
    },
    {
      future: [],
      past: [],
    },
  );
}

export function filterOnlyBookings(
  timeFrame: 'future' | 'past',
  bookings: $ReadOnlyArray<BookingsItem>,
) {
  const bookingGroups = groupBookings(bookings);

  if (timeFrame === 'future') {
    return bookingGroups.future;
  }
  return bookingGroups.past;
}

/**
 * Past booking is booking where every arrival date of every leg is in past
 * (considering local time).
 */
export function isPastBooking(singleBooking: BookingsItem) {
  const when = singleBooking.arrival.when;
  if (!when) {
    return true;
  }

  return (
    DateTime.fromJSDate(when.utc, {
      zone: 'utc',
    })
      .diffNow('milliseconds')
      .toObject().milliseconds < 0
  );
}

export function findUpcomingLeg(
  legs: Leg[],
  guarantee: ?$Values<typeof CoveredBy>,
) {
  return [...legs]
    .filter(leg => {
      const date = idx(leg, _ => _.arrival.when.utc);
      const isFuture = date
        ? DateTime.fromJSDate(date, { zone: 'utc' }).diffNow('milliseconds')
            .milliseconds > 0
        : false;

      return guarantee ? leg.guarantee === guarantee && isFuture : isFuture;
    })
    .sort((legA, legB) => {
      const dateA = idx(legA, _ => _.arrival.when.utc);
      const dateB = idx(legB, _ => _.arrival.when.utc);

      return dateA && dateB ? dateA - dateB : 0;
    })
    .shift();
}
