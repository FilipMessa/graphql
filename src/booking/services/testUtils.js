// @flow

export const createLeg = (milliseconds: number) => ({
  arrival: {
    when: {
      utc: new Date(milliseconds),
    },
  },
});
