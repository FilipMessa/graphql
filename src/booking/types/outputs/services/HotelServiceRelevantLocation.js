// @flow

import { GraphQLObjectType } from 'graphql';
import idx from 'idx';

import Location from '../../../../location/types/outputs/Location';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';
import HotelCityGraphqQLType from '../../../../hotel/types/outputs/HotelCity';

export default new GraphQLObjectType({
  name: 'HotelServiceRelevantLocation',
  fields: {
    location: {
      type: Location,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        return location;
      },
    },
    hotelCity: {
      type: HotelCityGraphqQLType,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader, locale }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        let hotelCities = [];
        const cityName = idx(location, _ => _.city.name);

        if (cityName) {
          hotelCities = await dataLoader.hotel.cities.loadByPrefix(
            cityName,
            locale.language,
          );
        }

        const lat = idx(location, _ => _.location.lat);
        const lng = idx(location, _ => _.location.lng);
        if (hotelCities.length === 0 && lat != null && lng != null) {
          /**
           * If we don't find any city quering by name, let's fall back to the coordinates of the airport
           * Note that this is most likely not going to be the destination city
           */
          hotelCities = await dataLoader.hotel.cities.loadByLatLng(lat, lng);
        }

        return hotelCities[0];
      },
    },
  },
});
