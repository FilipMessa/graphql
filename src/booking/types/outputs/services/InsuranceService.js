// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import GraphQLPassenger from '../Passenger';

export default new GraphQLObjectType({
  name: 'InsuranceService',
  fields: {
    passengers: {
      type: GraphQLList(GraphQLPassenger),
      description: 'Returns passengers allowed to buy a travel insurance.',
    },
  },
});
