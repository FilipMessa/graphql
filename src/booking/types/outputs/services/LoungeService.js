// @flow

import { GraphQLObjectType, GraphQLString, GraphQLList } from 'graphql';
import Location from '../../../../location/types/outputs/Location';
import LoungeWhiteLabelURLResolver from '../../../resolvers/LoungeWhitelabelURL';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';

type LoungeServiceInput = {|
  +iataCodes: Set<string>,
  +departureTime: Date,
|};

type LoungeServiceOutput = {|
  +iataCode: string,
  +departureTime: Date,
|};

export default new GraphQLObjectType({
  name: 'LoungeService',
  fields: {
    relevantAirports: {
      type: GraphQLList(
        new GraphQLObjectType({
          name: 'LoungeServiceRelevantAirports',
          fields: {
            whitelabelURL: {
              type: GraphQLString,
              resolve: async ({
                iataCode,
                departureTime,
              }: LoungeServiceOutput) => {
                return LoungeWhiteLabelURLResolver(iataCode, departureTime);
              },
            },

            location: {
              type: Location,
              resolve: (
                { iataCode }: LoungeServiceOutput,
                args: void,
                context: GraphqlContextType,
              ) => {
                return context.dataLoader.location.load(iataCode);
              },
            },
          },
        }),
      ),
      resolve: ({
        iataCodes,
        departureTime,
      }: LoungeServiceInput): $ReadOnlyArray<LoungeServiceOutput> => {
        return Array.from(iataCodes).map(iataCode => ({
          iataCode,
          departureTime,
        }));
      },
    },
  },
});
