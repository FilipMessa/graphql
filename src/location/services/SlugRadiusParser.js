// @flow

type SlugRadius = {|
  slug: string,
  radius: number,
|};
type GpsRadius = {|
  gps: { lat: number, lng: number },
  radius: number,
|};

export default function parse(slugRadius: string): SlugRadius | GpsRadius {
  const radius = getRadius(slugRadius);
  const slug = stripRadius(slugRadius, radius);

  // slug is not in lat-lng format
  if (slug.match(/^[0-9.-]+$/g) === null) {
    return { slug, radius };
  }

  const gps = getGps(slug);
  return { gps, radius };
}

function getRadius(slugRadius: string): number {
  const regex = RegExp('([0-9]+)km$', 'g');
  const parts: string[] = regex.exec(slugRadius);
  if (!Array.isArray(parts)) {
    throw new Error(`Error parsing radius. Input: ${slugRadius}`);
  }

  const radius = Number(parts[1]);
  if (Number.isNaN(radius) || !Number.isFinite(radius) || radius <= 0) {
    throw new Error(`Error parsing radius. Input: ${slugRadius}`);
  }
  return radius;
}

function getGps(slug: string): {| lat: number, lng: number |} {
  const { value: lat, slug: stripped } = stripCoordinate(slug);
  const { value: lng } = stripCoordinate(stripped);
  return { lat, lng };
}

function stripCoordinate(slug: string): {| value: number, slug: string |} {
  const parts = slug.match(/-?\d+(\.\d+)?/);
  if (!parts) {
    throw new Error(`Error parsing location. Input: ${slug}`);
  }
  const value = Number(parts[0]);

  const stripped = slug.substr(`${parts[0]}-`.length);
  return { value, slug: stripped };
}

function stripRadius(slugRadius: string, radius: number): string {
  const r = `-${radius}km`;
  const slug = slugRadius.substr(0, slugRadius.length - r.length);
  if (!slug.length)
    throw new Error(`Error parsing radius. Input: ${slugRadius}`);
  return slug;
}
