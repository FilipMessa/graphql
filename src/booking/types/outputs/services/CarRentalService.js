// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import CarRentalServiceRelevantCities, {
  type CarRentalServiceRelevantCitiesType,
} from './CarRentalServiceRelevantCities';
import { type GraphqlContextType } from '../../../../common/services/GraphqlContext';

export type CarRentalServiceType = Map<
  string,
  {|
    +pickup: Date,
    +dropoff: Date,
  |},
>;

export default new GraphQLObjectType({
  name: 'CarRentalService',
  fields: {
    relevantCities: {
      type: GraphQLList(CarRentalServiceRelevantCities),
      resolve: async (
        ancestor: CarRentalServiceType,
        args: {},
        context: GraphqlContextType,
      ): Promise<CarRentalServiceRelevantCitiesType[]> => {
        const locations = await Promise.all(
          Array.from(ancestor.keys()).map(IATA =>
            context.dataLoader.location.loadById(IATA),
          ),
        );

        return locations.map(location => {
          const ancestorValue = ancestor.get(location.locationId) || {
            pickup: new Date(),
            dropoff: new Date(),
          };

          return {
            location,
            pickup: ancestorValue.pickup,
            dropoff: ancestorValue.dropoff,
          };
        });
      },
    },
  },
});
