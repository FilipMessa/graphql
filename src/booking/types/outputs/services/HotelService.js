// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import HotelServiceRelevantLocation from './HotelServiceRelevantLocation';

type AncestorType = {|
  +relevantLocationCodes: $ReadOnlyArray<string>,
|};

export default new GraphQLObjectType({
  name: 'HotelService',
  fields: {
    relevantLocations: {
      type: GraphQLList(HotelServiceRelevantLocation),
      resolve: ({ relevantLocationCodes }: AncestorType) =>
        relevantLocationCodes,
    },
  },
});
