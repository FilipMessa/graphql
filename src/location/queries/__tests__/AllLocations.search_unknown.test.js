// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import config from '../../../../config/application';
import PragueDataset from '../../datasets/prague.json';
import UnknownLocationDataset from '../../datasets/unknown.json';

RestApiMock.onGet(
  config.restApiEndpoint.allLocations({
    term: 'unknown place',
    locale: 'en-US',
  }),
).replyWithData(UnknownLocationDataset);

RestApiMock.onGet(
  config.restApiEndpoint.allLocations({ term: 'Prague', locale: 'en-US' }),
).replyWithData(PragueDataset);

describe('all locations query', () => {
  it('should return successful searches', async () => {
    const query = `{
      A: allLocations(search: "Prague") {
        edges { node { name } }
      }
      B: allLocations(search: "unknown place") {
        edges { node { name } }
      }
    }`;
    expect(await graphql(query)).toMatchSnapshot();
  });

  it('should return successful searches - changed order', async () => {
    const query = `{
      A: allLocations(search: "unknown place") {
        edges { node { name } }
      }
      B: allLocations(search: "Prague") {
        edges { node { name } }
      }
    }`;
    expect(await graphql(query)).toMatchSnapshot();
  });
});
