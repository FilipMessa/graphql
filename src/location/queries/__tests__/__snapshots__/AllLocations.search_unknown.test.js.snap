// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all locations query should return successful searches - changed order 1`] = `
Object {
  "data": Object {
    "A": null,
    "B": Object {
      "edges": Array [
        Object {
          "node": Object {
            "name": "Prague",
          },
        },
        Object {
          "node": Object {
            "name": "Václav Havel Airport Prague",
          },
        },
        Object {
          "node": Object {
            "name": "Central Europe",
          },
        },
        Object {
          "node": Object {
            "name": "Europe",
          },
        },
      ],
    },
  },
  "errors": Array [
    [GraphQLError: Location has not been found.],
  ],
}
`;

exports[`all locations query should return successful searches 1`] = `
Object {
  "data": Object {
    "A": Object {
      "edges": Array [
        Object {
          "node": Object {
            "name": "Prague",
          },
        },
        Object {
          "node": Object {
            "name": "Václav Havel Airport Prague",
          },
        },
        Object {
          "node": Object {
            "name": "Central Europe",
          },
        },
        Object {
          "node": Object {
            "name": "Europe",
          },
        },
      ],
    },
    "B": null,
  },
  "errors": Array [
    [GraphQLError: Location has not been found.],
  ],
}
`;
